File: ocaml/lib/domain/point_transaction.ml
type transaction_id = TransactionId of int

type transaction_reason =
  | TaskCompletion of Task.task_id
  | RewardRedemption of Reward.reward_id
  | ManualAdjustment of string

type t = {
  id : transaction_id;
  user_id : User.user_id;
  amount : int;
  reason : transaction_reason;
  created_at : Ptime.t;
}

File: haskell/src/Domain/PointTransaction.hs
module Domain.PointTransaction where

import Data.Text (Text)
import Data.Time (UTCTime)
import Domain.Reward (RewardId)
import Domain.Task (TaskId)
import Domain.User (UserId)

newtype TransactionId = TransactionId Int deriving (Show, Eq)

data TransactionReason
  = TaskCompletion TaskId
  | RewardRedemption RewardId
  | ManualAdjustment Text
  deriving (Show, Eq)

data PointTransaction = PointTransaction
  { transactionId :: TransactionId,
    userId :: UserId,
    amount :: Int,
    reason :: TransactionReason,
    createdAt :: UTCTime
  }
  deriving (Show, Eq)
